### PR TITLE
Improve implementation of dunder equals

### PR DESCRIPTION
**Pull request by dmontagu**
Inspired by #3066, with tweaks for v2 and special handling of generic models to simplify possible edge cases.

Selected Reviewer: @samuelcolvin

### CODE REVIEW

**Comment by dmontagu** [comment]
please review

**Comment by adriangb** [comment]
The new behavior looks good to me. I think this is the sort of thing where we should just make the breaking change 👍

File: pydantic/main.py [hunk 1]
```diff
@@ -652,10 +652,27 @@ def _calculate_keys(
...
+        # only require their generic origin types to be equal, rather than exact type equality.
+        # This prevents headaches like MyGeneric(x=1) != MyGeneric[Any](x=1).
+        self_type = getattr(self, '__pydantic_generic_origin__', None) or self.__class__
+        other_type = getattr(other, '__pydantic_generic_origin__', None) or other.__class__
+
+        if self_type != other_type:
+            return False
+
+        if self.__dict__ != other.__dict__:
+            return False
```
Comment (dmontagu) [inline pydantic/main.py:668]: Previously I had this as a check of a tuple, but that was repeatably about 10% slower in the case where the types were different. The difference when the types were the same was negligible.

**Comment by samuelcolvin** [comment]
awesome, thank you.

**Pull request merged**